subsys/bluetooth/controller/ll_sw/isoal.c

```diff
@@ -0,0 +1,409 @@
+
+static struct
+{
+	isoal_alloc_state_t  sink_allocated[ISOAL_SINKS_MAX];
+	struct isoal_sink    sink_state[ISOAL_SINKS_MAX];
+} isoal_global;
+
+
+/**
+ * @brief Internal reset
+ * Zero-init entire ISO state
+ */
+static isoal_status_t _init_reset(void)
+{
+	memset(&isoal_global, 0, sizeof(isoal_global));
+	return ISOAL_STATUS_OK;
+}
+
+/**
+ * @brief  Initialize ISO-AL
+ */
+isoal_status_t isoal_init(void)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	err = _init_reset();
+	if (err) {
+		return err;
+	}
+
+	return err;
+}
+
+/** Clean up and reinitialize */
+isoal_status_t isoal_reset(void)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	err = _init_reset();
+	if (err) {
+		return err;
+	}
+
+	return err;
+}
+
+/**
+ * @brief Find free sink from statically-sized pool and allocate it
+ * @details Implemented as linear search since pool is very small
+ *
+ * @param hdl[out]  Handle to sink
+ * @return ISOAL_STATUS_OK if we could allocate; otherwise ISOAL_STATUS_ERR_SINK_ALLOC
+ */
+static isoal_status_t isoal_sink_allocate(isoal_sink_handle_t *hdl)
+{
+	isoal_sink_handle_t i;
+
+	/* Very small linear search to find first free */
+	for (i = 0; i < ISOAL_SINKS_MAX; i++) {
+		if (isoal_global.sink_allocated[i] == ISOAL_ALLOC_STATE_FREE) {
+			isoal_global.sink_allocated[i] = ISOAL_ALLOC_STATE_TAKEN;
+			*hdl = i;
+			return ISOAL_STATUS_OK;
+		}
+	}
+
+	return ISOAL_STATUS_ERR_SINK_ALLOC; /* All entries were taken */
+}
+
+/**
+ * @brief Mark a sink as being free to allocate again
+ * @param hdl[in]  Handle to sink
+ */
+static void isoal_sink_deallocate(isoal_sink_handle_t hdl)
+{
+	isoal_global.sink_allocated[hdl] = ISOAL_ALLOC_STATE_FREE;
+}
+
+/**
+ * @brief Create a new sink
+ *
+ * @param hdl[out]          Handle to new sink
+ * @param sdu_alloc[in]   Callback of sdu allocator
+ * @param sdu_emit[in]    Callback of sdu emitter
+ * @return ISOAL_STATUS_OK if we could create a new sink; otherwise ISOAL_STATUS_ERR_SINK_ALLOC
+ */
+isoal_status_t isoal_sink_create(
+	isoal_sink_handle_t      *hdl,
+	isoal_sink_sdu_alloc_cb  sdu_alloc,
+	isoal_sink_sdu_emit_cb   sdu_emit)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	/* Allocate a new sink */
+	err = isoal_sink_allocate(hdl);
+	if (err) {
+		return err;
+	}
+
+	/* Remember the platform-specific callbacks */
+	isoal_global.sink_state[*hdl].session.sdu_alloc = sdu_alloc;
+	isoal_global.sink_state[*hdl].session.sdu_emit  = sdu_emit;
+
+	/* Initialize running seq number to zero */
+	isoal_global.sink_state[*hdl].session.seqn = 0;
+
+	return err;
+}
+
+/**
+ * @brief Configure a sink
+ * @param hdl[in]     Handle to new sink
+ * @param config[in]  Params
+ */
+isoal_status_t isoal_sink_configure(
+	isoal_sink_handle_t hdl,
+	const struct isoal_sink_config *param
+)
+{
+	isoal_status_t err = ISOAL_STATUS_OK;
+
+	/* Copy parameters so we know they will survive for the whole session */
+	memcpy(&isoal_global.sink_state[hdl].session.param, param,
+	       sizeof(struct isoal_sink_config));
```

Comment:
We can avoid this copy, moving it the responsibility to the caller. That would be a more flexible API, as the caller doesn't have to first build up an identical struct -- it may as well just directly write into the dest param struct.

I.e. we can just return a pointer to the param struct. And make sure the param struct is in public header.